### ISSUE TITLE
Memory leak when training models

### ISSUE THREAD

**Issue by HayeonLee** [google-research/meta-dataset#28]
When training models you provided such as the prototypical net or maml, I found the memory increase until 60~90G and ResourceExhausted error with OOM occurred.

I guess tfrecords are related with the memory leak. 
I applied tf.reset_default_graph() or tf.keras.backend.clear_session()  to the code, but it looks not working.  
Could you help me to fix this problem?
![image](https://user-images.githubusercontent.com/19198395/64925558-3ec90480-d82d-11e9-9c0d-5e7ae86639f5.png)
![image](https://user-images.githubusercontent.com/19198395/64925561-54d6c500-d82d-11e9-86c2-bb004158a7ff.png)

**Comment by gabrielhuang** [comment]
I had the same type of error, and the temporary fix that works for me is to reduce the shuffle_buffer_size:
```
--gin-bindings="DataConfig.shuffle_buffer_size=300"
```

**Comment by gabrielhuang** [comment]
The other thing is to increase the limit of open files
```bash
ulimit -n 100000
python [...]
```

**Comment by raghavgoyal14** [comment]
@gabrielhuang: as a follow up do you know a workaround for users who cannot increase the `ulimit` as non-root users?

**Comment by gabrielhuang** [comment]
I could increase to 100,000 as a non root user just by running uimit - n
100000 in the terminal.
Otherwise I don't know the code well enough to comment. good luck!

**Comment by lamblin** [comment]
Depending on your system, you may not need to be root to increase the `ulimit` (although you cannot re-raise it after you reduce it I think).

As for workarounds, the main issue is that we store each class as a file, to be able to assemble episodes by mixing examples with changing proportions. So with > 1000 classes, you'd need that many files.
The only (unsatisfying) solutions I can see are:
- Pre-generate a long stream of episodes from a machine you control, then every worker can just read that one stream sequentially (or slightly shuffled). For 50k episodes, it probably brings you in the terabyte range.
- Load everything in memory. You'd need at least 210 GB just to store the data, but you might be able to get rid of the shuffle queue and just access the elements directly.
- Something similar, but split each class in several files, and load only one of them in memory at a time, replacing them regularly. I have no idea how to integrate that to the tf.data pipeline.

**Comment by raghavgoyal14** [comment]
@lamblin  Thanks for the response!

Naive question: Why do you need to open all the classes' files for generating an episode? IIRC classes are sampled for an episode so you'd need to open only the sampled classes' files relevant for the episode and then close it.


Observation: I was able to run the code using a `ulimit -n` of `4096` as opposed to what I had before (`1024`)
And after looking at the number of classes, the highest number of classes is 1623 from the Omniglot dataset. Does that mean that I only need to comfortably set a `ulimit -n` of more than 1623 to run the code (including all my fileIO activities too)? Do you think this might be the reason that I got the error with `ulimit -n` of `1024`?

Thanks again!

**Comment by lamblin** [comment]
The tf.data pipeline is taking care of file opening / reading / closing, so I'm not exactly sure what is going on under the hood. .tfrecords files are meant to be read sequentially, I think, so it could be quite inefficient to open / seek / close the files for each episode. It's possible that there are optimizations already, like short files being read at once and not left open.

If you train on multiple datasets (for instance in the "all" setting), then files from all datasets could be open at the same time, not only Omniglot.

An upper bound on the number of open files in the "all" setting would be:
- 3144 tfrecords files for train
- 158 tfrecords files for valid (the default is to validate only on ImageNet)
- 8 dataset_spec.json + 1 num_leaf_images.json
- 3 files for each checkpoint + 1 for summaries

So > 3300, not including the source files, interpreter, etc. So a limit of 4096 sounds reasonable in your case. If you include more datasets for validation, it could be 441 more, which should still be OK.

**Comment by raghavgoyal14** [comment]
@lamblin: This helps. Thank you very much for the detailed answer!
